src/Plugins/ComplexCore/src/ComplexCore/Filters/RotateSampleRefFrame.cpp

```diff
@@ -0,0 +1,540 @@
+#include "RotateSampleRefFrame.hpp"
+
+#include "complex/Common/ComplexRange3D.hpp"
+#include "complex/Common/Numbers.hpp"
+#include "complex/Common/TypeTraits.hpp"
+#include "complex/DataStructure/Geometry/ImageGeom.hpp"
+#include "complex/Filter/Actions/CreateArrayAction.hpp"
+#include "complex/Filter/Actions/CreateImageGeometryAction.hpp"
+#include "complex/Parameters/ChoicesParameter.hpp"
+#include "complex/Parameters/DataGroupCreationParameter.hpp"
+#include "complex/Parameters/DynamicTableParameter.hpp"
+#include "complex/Parameters/GeometrySelectionParameter.hpp"
+#include "complex/Parameters/MultiArraySelectionParameter.hpp"
+#include "complex/Parameters/NumberParameter.hpp"
+#include "complex/Parameters/VectorParameter.hpp"
+#include "complex/Utilities/FilterUtilities.hpp"
+#include "complex/Utilities/Math/MatrixMath.hpp"
+#include "complex/Utilities/ParallelData3DAlgorithm.hpp"
+
+#include <Eigen/Dense>
+
+#include <fmt/core.h>
+
+#include <nonstd/span.hpp>
+
+#include <algorithm>
+#include <array>
+#include <cmath>
+#include <limits>
+#include <stdexcept>
+
+using namespace complex;
+
+namespace
+{
+using RotationRepresentation = RotateSampleRefFrame::RotationRepresentation;
+using Matrix3fR = Eigen::Matrix<float32, 3, 3, Eigen::RowMajor>;
+
+constexpr float32 k_Threshold = 0.01f;
+
+const Eigen::Vector3f k_XAxis = Eigen::Vector3f::UnitX();
+const Eigen::Vector3f k_YAxis = Eigen::Vector3f::UnitY();
+const Eigen::Vector3f k_ZAxis = Eigen::Vector3f::UnitZ();
+
+struct RotateArgs
+{
+  int64 xp = 0;
+  int64 yp = 0;
+  int64 zp = 0;
+  float32 xRes = 0.0f;
+  float32 yRes = 0.0f;
+  float32 zRes = 0.0f;
+  int64 xpNew = 0;
+  int64 ypNew = 0;
+  int64 zpNew = 0;
+  float32 xResNew = 0.0f;
+  float32 yResNew = 0.0f;
+  float32 zResNew = 0.0f;
+  float32 xMinNew = 0.0f;
+  float32 yMinNew = 0.0f;
+  float32 zMinNew = 0.0f;
+};
+
+class SampleRefFrameRotator
+{
+public:
+  SampleRefFrameRotator(nonstd::span<int64> newIndices, const RotateArgs& args, const Matrix3fR& rotationMatrix, bool sliceBySlice)
+  : m_NewIndices(newIndices)
+  , m_SliceBySlice(sliceBySlice)
+  , m_Params(args)
+  {
+    // We have to inline the 3x3 Maxtrix transpose here because of the "const" nature of the 'convert' function
```

Comment:
// We have to inline the 3x3 Matrix transpose here because of the "const" nature of the 'convert' function